rsciio/topspin/_api.py

```diff
@@ -0,0 +1,432 @@
+            load_single_file = True
+
+    # Generate task list by searching .app5 for GUID's
+    task_list = []
+    group_names = [x for x in h5_file.keys()]
+    if np.isin("Metadata", group_names):  # single experiment file.
+        for name in h5_file.keys():
+            if looks_like_a_guid(name):
+                task_list.append([name, "Metadata"])
+    else:  # Multi-experimental file.
+        for top_name in h5_file.keys():
+            if looks_like_a_guid(top_name):
+                for name in h5_file[top_name].keys():
+                    if looks_like_a_guid(name):
+                        address = top_name + "/" + name
+                        meta = top_name + "/" + "Metadata"
+                        task_list.append([address, meta])
+    # prune list for single-file query
+    if load_single_file:
+        task_list = [x for x in task_list if x[0] == subset]
+
+    # Parse each dataset's metadata
+    datasets_list = []
+    for address, meta_loc in task_list:
+        name = address.split("/")[-1]
+        xml_str = h5_file[meta_loc][()].decode()
+        original_meta = _parse_app5_xml(xml_str)
+        hspy_meta = _parse_hspy_meta(original_meta, filename)
+        if isinstance(h5_file[address], h5py.Group):
+            # 4D-STEM
+            h5_grp = h5_file[address]
+            axes = _get_4D_axes(h5_grp)
+            hspy_meta["Signal"]["signal_type"] = "electron_diffraction"
+        else:
+            # 2D composite image
+            axes = _get_2D_axes(xml_str, name)
+            shape = h5_file[address].shape
+            axes[0]["size"] = shape[0]
+            axes[1]["size"] = shape[1]
+            hspy_meta["Signal"]["signal_type"] = "STEM"
+        datasets_list.append(
+            {
+                "axes": axes,
+                "metadata": hspy_meta,
+                "original_metadata": original_meta,
+            }
+        )
+
+    if dryrun:
+        filename = str(filename)
+        message = "The following would be imported from " + filename + ":"
+        for i, ds_dict in enumerate(datasets_list):
+            shape = [x["size"] for x in ds_dict["axes"]]
+            guid = task_list[i][0]
+            name = ds_dict["original_metadata"]["Title"]
+            message += "\n   - {}, {}\n      {}".format(name, shape, guid)
+        sys.stdout.write(message)
+        return []
+
+    for i, ds_dict in enumerate(datasets_list):
+        shape = [x["size"] for x in ds_dict["axes"]]
+        address = task_list[i][0]
+        if len(shape) == 2:
+            # this progress bar is really just a "1-of-1", but it still
+            # reports load time and confirms loading succeeded
+            for dummy_iter in tqdm(
+                [1],
+                desc="Loading {}: ".format(address),
+                disable=not show_progressbar,
+            ):
+                datasets_list[i]["data"] = np.asanyarray(h5_file[address])
+
+        elif len(shape) == 4:
```

Comment:
https://github.com/hyperspy/rosettasciio/pull/454/files#r2552811344
@ericpre : 
> Do you have a file with a dataset with 3 dimension? Would it be possible to get one? If so, it can be done in a separate PR.



I don't think so.

Internally, the SPED datasets are stored like a vector of 2D diffraction patterns, (so, pseudo-3d), then reshaped into 4D during loading. Thus, I think even a 1x400x256x256 map would show up as a 4D signal.

Reply:
fixed.